### DISCUSSION TITLE
Branching

### DISCUSSION THREAD

**Discussion by wormstein** [EdNekebno/DGTCentaurMods#59] [General]
Hey,
Just proposing, in the end you decide because it's your repo.

Could you start working on a branch that we merge in master once we have nailed features.

I know you're not very familiar with GitHub but I am happy to help.

The reason I say is because we explore multiple things and we should stabilize master branch.

Later when I get home I can post some basic commands and explain some more stuff.

+A

**Comment by EdNekebno** [comment]
Strangely I was l trying to teach myself stuff about branching earlier. There is almost certainly some best practice here that I am going to completely ignore :) But we're essentially talking about how much the master branch gets updated.

I take the view, I think, that once we have something that we consider 'production ready' then we should absolutely have branches for each feature and issue and quite possibly a development branch to merge into and test before modifying master.

But until that the fastest way to get to that point is to push often so that we see the code changes early and any potential problems early. Then the word 'master' is probably a bit of a misnomer. It's really a development branch.

So I go with - absolutely - when we have releases and people are using it (who aren't linux geniuses) then we stabilise it. And in the meantime I try to learn what the heck I'm doing :)

For me I consider that what we have + lichess being converted to run with gamanager.py + at least a basic PGN export done is a worthwhile feature set to consider a first version. So we're not a million miles away. If we can do that and bug fix it...then I promise proper grown up github branching, etc :)

**Comment by wormstein** [comment]
Sounds good. I think I can bring something by the end of the weekend.
I have a second RPi here ready to break on testing the deployment :))

**Comment by EdNekebno** [comment]
I was looking to buy another pi zero w today.....but everywhere I looked online didn't have any. Not sure if it's a general supply problem or if it's a problem being in a small country.

**Comment by wormstein** [comment]
We now have two major branches to avoid breaking the master.

- build - this will hold changes related to build/install process that needs testing. Pushing changes here will allow other users to get them and test, Upstream repo is master.
- dev - holds any new features to the mod. Same idea here.

Should we have a QA branch which will merge all the changes from the orher ones and be the final step for testing?

Cheers

**Comment by cyberluke** [comment]
In this case build is like a testing branch. Dev is development, but it should contain changes from build branch as well. So the flow should be: Dev > Build > Master. Is that correct?

Because I develop new feature and need to have build scripts up to date as well. There should be a one way flow otherwise there is a risk of merge and dependency hell.

**Comment by wormstein** [comment]
Not really. We don't have to build to test new features unless changes are related to build process.

If your feature requires adjustements in building the installer, you can push those in dev so we know they're related.l to your new feature.

Those will end up in master then merged in build with other major updates to build, like update handling, etc, and then after testing the build process against the master branch they merge back there.

Releases will only be for end user. We have a manual setup where we can easily test the code without having to rebuild.
